File: HerdMe/Services/ArtisanCommandRunner.swift
import Foundation

struct ArtisanCommandPreset: Identifiable, Hashable, Sendable {
    let id: String
    let title: String
    let arguments: [String]
    let timeout: TimeInterval

    static let all: [ArtisanCommandPreset] = [
        .init(id: "route-list", title: "Route List", arguments: ["route:list", "--no-interaction"], timeout: 5 * 60),
        .init(id: "migrate-status", title: "Migration Status", arguments: ["migrate:status", "--no-interaction"], timeout: 5 * 60),
        .init(id: "migrate", title: "Migrate", arguments: ["migrate", "--no-interaction"], timeout: 15 * 60),
        .init(id: "queue-work", title: "Queue Worker", arguments: ["queue:work", "--no-interaction"], timeout: 24 * 60 * 60),
        .init(id: "custom", title: "Custom", arguments: [], timeout: 15 * 60)
    ]
}

enum ArtisanCommandError: LocalizedError, Equatable {
    case empty
    case malformedQuotes
    case invalidCharacter
    case tooLong
    case tooManyArguments
    case invalidCommand

    var errorDescription: String? {
        switch self {
        case .empty:
            String(localized: "Enter an Artisan command.")
        case .malformedQuotes:
            String(localized: "The Artisan command contains an unfinished quote.")
        case .invalidCharacter:
            String(localized: "The Artisan command contains an unsupported control character.")
        case .tooLong:
            String(localized: "Artisan commands are limited to 4 KB and 512 bytes per argument.")
        case .tooManyArguments:
            String(localized: "Artisan commands are limited to 32 arguments.")
        case .invalidCommand:
            String(localized: "Enter an Artisan command such as route:list, without php or shell syntax.")
        }
    }
}

enum ArtisanCommandParser {
    private static let maximumCommandBytes = 4 * 1_024
    private static let maximumArgumentBytes = 512
    private static let maximumArguments = 32

    static func arguments(presetID: String, customCommand: String) throws -> (arguments: [String], timeout: TimeInterval) {
        guard let preset = ArtisanCommandPreset.all.first(where: { $0.id == presetID }) else {
            throw ArtisanCommandError.invalidCommand
        }
        if preset.id == "custom" {
            return (try parse(customCommand), preset.timeout)
        }
        return (preset.arguments, preset.timeout)
    }

    static func parse(_ command: String) throws -> [String] {
        let trimmed = command.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw ArtisanCommandError.empty }
        guard trimmed.utf8.count <= maximumCommandBytes else { throw ArtisanCommandError.tooLong }
        guard
            !trimmed.unicodeScalars.contains(where: { scalar in
                scalar.value == 0 || scalar.value < 32 || scalar.value == 127
            })
        else {
            throw ArtisanCommandError.invalidCharacter
        }

        enum Quote: Equatable {
            case single
            case double
        }

        var quote: Quote?
        var isEscaping = false
        var token = ""
        var tokenStarted = false
        var arguments: [String] = []

        func appendToken() throws {
            guard token.utf8.count <= maximumArgumentBytes else { throw ArtisanCommandError.tooLong }
            arguments.append(token)
            guard arguments.count <= maximumArguments else { throw ArtisanCommandError.tooManyArguments }
            token = ""
            tokenStarted = false
        }

        for character in trimmed {
            if isEscaping {
                token.append(character)
                tokenStarted = true
                isEscaping = false
                continue
            }

            if character == "\\", quote != .single {
                isEscaping = true
                tokenStarted = true
                continue
            }

            if character == "'", quote != .double {
                quote = quote == .single ? nil : .single
                tokenStarted = true
                continue
            }

            if character == "\"", quote != .single {
                quote = quote == .double ? nil : .double
                tokenStarted = true
                continue
            }

            if character.isWhitespace, quote == nil {
                if tokenStarted { try appendToken() }
                continue
            }

            token.append(character)
            tokenStarted = true
        }

        guard !isEscaping, quote == nil else { throw ArtisanCommandError.malformedQuotes }
        if tokenStarted { try appendToken() }
        if arguments.first == "artisan" { arguments.removeFirst() }
        guard let commandName = arguments.first,
            !commandName.isEmpty,
            !commandName.hasPrefix("-"),
            commandName != "php"
        else {
            throw ArtisanCommandError.invalidCommand
        }
        return arguments
    }
}

struct ArtisanInvocation: Sendable {
    let phpExecutable: URL
    let projectDirectory: URL
    let arguments: [String]
    let environment: [String: String]
    let timeout: TimeInterval
}

struct ArtisanCommandResult: Sendable {
    let status: Int32
    let output: String
}

final class ArtisanCancellation: @unchecked Sendable {
    private let lock = NSLock()
    private var cancelled = false

    var isCancelled: Bool {
        lock.lock()
        defer { lock.unlock() }
        return cancelled
    }

    func cancel() {
        lock.lock()
        cancelled = true
        lock.unlock()
    }
}

enum ArtisanCommandRunner {
    static func run(
        _ invocation: ArtisanInvocation,
        cancellation: ArtisanCancellation,
        outputReceived: @escaping @Sendable (Data) -> Void
    ) async throws -> ArtisanCommandResult {
        let task = Task.detached(priority: .userInitiated) {
            let result = try ProcessRunner.run(
                invocation.phpExecutable,
                arguments: ["artisan"] + invocation.arguments,
                currentDirectory: invocation.projectDirectory,
                environment: invocation.environment,
                timeout: invocation.timeout,
                cancellationRequested: { cancellation.isCancelled },
                outputReceived: outputReceived
            )
            return ArtisanCommandResult(status: result.status, output: result.output)
        }
        return try await withTaskCancellationHandler {
            try await task.value
        } onCancel: {
            cancellation.cancel()
            task.cancel()
        }
    }
}

File: Windows/HerdMe.Windows/Services/ArtisanCommandRunner.cs
using System.Diagnostics;
using System.Text;

namespace HerdMe.Windows.Services;

public sealed record ArtisanCommandPreset(
    string Id,
    string Title,
    IReadOnlyList<string> Arguments,
    TimeSpan Timeout
);

public sealed record ArtisanCommandSpec(
    IReadOnlyList<string> Arguments,
    TimeSpan Timeout
);

public sealed record ArtisanCommandResult(int ExitCode, string Output);

public static class ArtisanCommandCatalog
{
    private const int MaximumCommandBytes = 4 * 1_024;
    private const int MaximumArgumentBytes = 512;
    private const int MaximumArguments = 32;

    public static IReadOnlyList<ArtisanCommandPreset> Presets { get; } = Array.AsReadOnly(
        new[]
        {
            new ArtisanCommandPreset("route-list", "Route List", ["route:list", "--no-interaction"], TimeSpan.FromMinutes(5)),
            new ArtisanCommandPreset("migrate-status", "Migration Status", ["migrate:status", "--no-interaction"], TimeSpan.FromMinutes(5)),
            new ArtisanCommandPreset("migrate", "Migrate", ["migrate", "--no-interaction"], TimeSpan.FromMinutes(15)),
            new ArtisanCommandPreset("queue-work", "Queue Worker", ["queue:work", "--no-interaction"], TimeSpan.FromHours(24)),
            new ArtisanCommandPreset("custom", "Custom", [], TimeSpan.FromMinutes(15))
        }
    );

    public static ArtisanCommandSpec Resolve(string presetId, string customCommand)
    {
        var preset = Presets.SingleOrDefault(item => item.Id == presetId)
            ?? throw new ArgumentException("Choose a supported Artisan command.", nameof(presetId));
        return preset.Id == "custom"
            ? new ArtisanCommandSpec(Parse(customCommand), preset.Timeout)
            : new ArtisanCommandSpec(preset.Arguments, preset.Timeout);
    }

    public static IReadOnlyList<string> Parse(string command)
    {
        var trimmed = command.Trim();
        if (trimmed.Length == 0) throw new ArgumentException("Enter an Artisan command.", nameof(command));
        if (Encoding.UTF8.GetByteCount(trimmed) > MaximumCommandBytes)
        {
            throw new ArgumentException("Artisan commands are limited to 4 KB.", nameof(command));
        }
        if (trimmed.Any(char.IsControl))
        {
            throw new ArgumentException(
                "The Artisan command contains an unsupported control character.",
                nameof(command)
            );
        }

        var arguments = new List<string>();
        var token = new StringBuilder();
        var quote = '\0';
        var escaping = false;
        var tokenStarted = false;

        void AppendToken()
        {
            var value = token.ToString();
            if (Encoding.UTF8.GetByteCount(value) > MaximumArgumentBytes)
            {
                throw new ArgumentException(
                    "Artisan arguments are limited to 512 bytes.",
                    nameof(command)
                );
            }
            arguments.Add(value);
            if (arguments.Count > MaximumArguments)
            {
                throw new ArgumentException(
                    "Artisan commands are limited to 32 arguments.",
                    nameof(command)
                );
            }
            token.Clear();
            tokenStarted = false;
        }

        foreach (var character in trimmed)
        {
            if (escaping)
            {
                token.Append(character);
                tokenStarted = true;
                escaping = false;
                continue;
            }
            if (character == '\\' && quote != '\'')
            {
                escaping = true;
                tokenStarted = true;
                continue;
            }
            if (character == '\'' && quote != '"')
            {
                quote = quote == '\'' ? '\0' : '\'';
                tokenStarted = true;
                continue;
            }
            if (character == '"' && quote != '\'')
            {
                quote = quote == '"' ? '\0' : '"';
                tokenStarted = true;
                continue;
            }
            if (char.IsWhiteSpace(character) && quote == '\0')
            {
                if (tokenStarted) AppendToken();
                continue;
            }
            token.Append(character);
            tokenStarted = true;
        }

        if (escaping || quote != '\0')
        {
            throw new ArgumentException(
                "The Artisan command contains an unfinished quote.",
                nameof(command)
            );
        }
        if (tokenStarted) AppendToken();
        if (arguments.FirstOrDefault() == "artisan") arguments.RemoveAt(0);
        if (arguments.Count == 0
            || arguments[0].Length == 0
            || arguments[0].StartsWith("-", StringComparison.Ordinal)
            || arguments[0] == "php")
        {
            throw new ArgumentException(
                "Enter an Artisan command such as route:list, without php or shell syntax.",
                nameof(command)
            );
        }
        return arguments;
    }
}

public static class ArtisanCommandRunner
{
    private const int MaximumCapturedCharacters = 1 * 1_024 * 1_024;

    public static async Task<ArtisanCommandResult> RunAsync(
        string phpExecutable,
        string projectDirectory,
        IReadOnlyList<string> arguments,
        IReadOnlyDictionary<string, string> environment,
        TimeSpan timeout,
        IProgress<string>? outputProgress = null,
        CancellationToken cancellationToken = default
    )
    {
        var projectPath = Path.GetFullPath(projectDirectory);
        if (!File.Exists(phpExecutable))
        {
            throw new FileNotFoundException("Install the selected HerdMe PHP runtime first.", phpExecutable);
        }
        if (!Directory.Exists(projectPath) || !File.Exists(Path.Combine(projectPath, "artisan")))
        {
            throw new InvalidOperationException(
                "Artisan is available only for Laravel projects with an artisan executable."
            );
        }
        if (arguments.Count == 0 || timeout <= TimeSpan.Zero)
        {
            throw new ArgumentException("The Artisan invocation is incomplete.", nameof(arguments));
        }

        var startInfo = new ProcessStartInfo
        {
            FileName = phpExecutable,
            WorkingDirectory = projectPath,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
            StandardOutputEncoding = Encoding.UTF8,
            StandardErrorEncoding = Encoding.UTF8
        };
        startInfo.ArgumentList.Add("artisan");
        foreach (var argument in arguments) startInfo.ArgumentList.Add(argument);
        foreach (var variable in environment) startInfo.Environment[variable.Key] = variable.Value;

        using var process = Process.Start(startInfo)
            ?? throw new InvalidOperationException("The selected PHP runtime could not start Artisan.");
        var capture = new BoundedOutputCapture(MaximumCapturedCharacters);
        var standardOutput = PumpAsync(process.StandardOutput, capture, outputProgress);
        var standardError = PumpAsync(process.StandardError, capture, outputProgress);
        using var timeoutCancellation = new CancellationTokenSource(timeout);
        using var linkedCancellation = CancellationTokenSource.CreateLinkedTokenSource(
            cancellationToken,
            timeoutCancellation.Token
        );
        try
        {
            await process.WaitForExitAsync(linkedCancellation.Token);
        }
        catch (OperationCanceledException) when (linkedCancellation.IsCancellationRequested)
        {
            try
            {
                if (!process.HasExited) process.Kill(entireProcessTree: true);
            }
            catch (InvalidOperationException)
            {
                // The process exited while cancellation was being delivered.
            }
            await process.WaitForExitAsync(CancellationToken.None);
            await Task.WhenAll(standardOutput, standardError);
            cancellationToken.ThrowIfCancellationRequested();
            throw new TimeoutException(
                $"The Artisan command did not finish within {timeout.TotalMinutes:0} minutes."
            );
        }
        await Task.WhenAll(standardOutput, standardError);
        return new ArtisanCommandResult(process.ExitCode, capture.Value);
    }

    private static async Task PumpAsync(
        StreamReader reader,
        BoundedOutputCapture capture,
        IProgress<string>? outputProgress
    )
    {
        var buffer = new char[4_096];
        while (true)
        {
            var count = await reader.ReadAsync(buffer.AsMemory());
            if (count == 0) return;
            var chunk = new string(buffer, 0, count);
            capture.Append(chunk);
            outputProgress?.Report(chunk);
        }
    }

    private sealed class BoundedOutputCapture(int maximumCharacters)
    {
        private readonly object sync = new();
        private readonly StringBuilder output = new();

        public string Value
        {
            get
            {
                lock (sync) return output.ToString();
            }
        }

        public void Append(string value)
        {
            lock (sync)
            {
                output.Append(value);
                if (output.Length > maximumCharacters)
                {
                    output.Remove(0, output.Length - maximumCharacters);
                }
            }
        }
    }
}
